[FEATURE] [Back] Compléter et rendre plus pratique la validation des locales (PIX-18904)

lego-technix opened the pull request:
## 🔆 Problème

1. Les routes suivantes n’ont qu’une validation très incomplète des paramètres passés et, pour les 2 premières routes, aucune validation de la `locale` : 
  * `POST` `/api/users`
  * `PATCH` `/api/users`
  * `PATCH` `/api/admin/users`
2. La validation de la `locale` d’un utilisateur, réalisée à plusieurs endroits, avec la fonction `getCanonicalLocale` et les erreurs `LocaleNotSupportedError` et `LocaleFormatError` n’est ni claire ni pratique.

## ⛱️ Proposition

1. Ajouter de la validation à tous les points d’entrée de l’API : 
   * validation Joi sur tous les attributs des routes concernées
   * validation du cookie `locale` au niveau des routes concernées
2. Ajouter une fonction `localeService.getNearestSupportedLocale` pour s'assurer que la locale transmise est une locale supportée par Pix

Cette PR supprime les erreurs spécifiques `LocaleNotSupportedError` et `LocaleFormatError` car on ne veut pas faire d’actions particulières sur ces erreurs et donc on n’a pas besoin de définir des types d’erreur particuliers pour ces erreurs. 

## 🌊 Remarques

Grâce à cette PR on note que le passage des paramètres aux routes concernées est réalisé en envoyant beaucoup trop d’informations 🤢 L’avantage de mettre en place avec cette PR une validation de manière complète (sans le `allowUnknown: true`) est que cela met évidence qu‘il faut reprendre l’utilisation de cette API par les frontaux pour n’envoyer et n’autoriser que les informations nécessaires.

## 🏄 Pour tester

Tester les actions utilisant les routes suivantes : 
  * `POST` `/api/users`
  * `PATCH` `/api/users`
  * `PATCH` `/api/admin/users`
  
C'est à dire concrètement : 
* Sur Pix App `.pix.fr` créer un utilisateur
* Sur Pix App `.pix.org` créer un utilisateur
* Sur Pix App (`.pix.fr` ou `.pix.org` indifférent) passer un parcours simplifié (utilisateur anonyme) et à l’issue créer un compte
  * https://app-pr13053.review.pix.org/campagnes/AUTOCOUR1
* Sur Pix Orga `.pix.fr`
  * envoyer une invitation
  * accepter cette invitation en créant un nouveau compte
* Sur Pix Orga `.pix.org`
  * envoyer une invitation
  * accepter cette invitation en créant un nouveau compte
* Sur Pix Certif `.pix.fr`
  * envoyer une invitation
  * accepter cette invitation en créant un nouveau compte
* Sur Pix Certif `.pix.org`
  * envoyer une invitation
  * accepter cette invitation en créant un nouveau compte
* Sur Pix Admin `.pix.fr`
  * modifier les informations d'un compte utilisateur.

pix-bot-github commented:
Choisir les applications à déployer :

- [x] [API](https://api-pr13053.review.pix.fr/api/) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr13053/environment) <!-- pix-api-review -->
- [x] [App (.fr)](https://app-pr13053.review.pix.fr) – [App (.org)](https://app-pr13053.review.pix.org) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-app-review-pr13053/environment) <!-- pix-app-review -->
- [x] [Orga (.fr)](https://orga-pr13053.review.pix.fr) – [Orga (.org)](https://orga-pr13053.review.pix.org) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-orga-review-pr13053/environment) <!-- pix-orga-review -->
- [x] [Certif (.fr)](https://certif-pr13053.review.pix.fr) – [Certif (.org)](https://certif-pr13053.review.pix.org) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-certif-review-pr13053/environment) <!-- pix-certif-review -->
- [ ] [Junior](https://junior-pr13053.review.pix.fr) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-junior-review-pr13053/environment) <!-- pix-junior-review -->
- [x] [Admin](https://admin-pr13053.review.pix.fr) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-admin-review-pr13053/environment) <!-- pix-admin-review -->
- [ ] [API MaDDo](https://pix-api-maddo-review-pr13053.osc-fr1.scalingo.io/api/) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-maddo-review-pr13053/environment) <!-- pix-api-maddo-review -->
- [ ] [Audit Logger](https://pix-audit-logger-review-pr13053.osc-fr1.scalingo.io/api/) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-audit-logger-review-pr13053/environment) <!-- pix-audit-logger-review -->

> [!IMPORTANT]
> N'oubliez pas de déployer l'API pour pouvoir accéder aux fronts et/ou à l’API MaDDo.

lego-technix committed:
test(api): more efficient way to deal with the localeService

lego-technix committed:
test(api): fix a typo in the API route in a test description

lego-technix committed:
feat(api): add complete validation on POST /api/users

lego-technix committed:
refactor(api): rather import the whole localeService

lego-technix committed:
test(api): remove duplicate tests for POST /api/users

lego-technix committed:
feat(api): add complete validation on PATCH /api/users

AndreiaPena commented:
✨ SAV Hera bonjour ✨

Nous avons clôturé puis ré-ouvert votre PR afin de passer sur le fonctionnement Hera.

Les apps Scalingo de votre pull request ont été supprimées automatiquement.
Pour y accéder, cochez les apps que vous souhaitez déployer sur le commentaire de pix-bot-github !

✨ Passez une bonne expérience avec Hera ✨

lego-technix committed:
feat(api): add complete validation on PATCH /api/admin/users/{id}

The actual passed parameters are nonetheless very questionable and this surely
needs to be dealt with later.

lego-technix committed:
test(api): add some missing databaseBuilder.commit()

lego-technix committed:
feat(api): add new function localeService.getNearestSupportedLocale

lego-technix committed:
refactor(api): leverage new function localeService.assertIsCanonicalLocaleName for clearer-without-side-effect validation

lego-technix committed:
refactor(api): remove now useless LocaleNotSupportedError

lego-technix committed:
refactor(api): remove now useless LocaleFormatError

lego-technix committed:
feat(api): add a missing locale cookie validation in OIDC createUser

lego-technix committed:
test(api): remove some Route | Admin | User useless tests

bpetetot commented on api/src/identity-access-management/application/user/user.route.js:
```diff
@@ -164,19 +164,33 @@ export const userRoutes = [
+              locale: Joi.string()
+                .valid(...localeService.getSupportedLocales())
+                .allow(null),
+              email: Joi.string().allow(null),
+              username: Joi.string().allow(null),
+              password: Joi.string().allow(null),
+              cgu: Joi.boolean().allow(null),
+              'anonymous-user-token': Joi.string().allow(null),
```
Les attributs `email`, `password` et `cgu` étaient required avant, c'est bizarre qu'elle soient maintenant nullable.

lego-technix replied:
Bien vu. Ce point a été corrigé.

J’ai mis un TODO pour améliorer la situation actuelle :
```
path: '/api/users/{id}', // TODO: Rename this route to reflect it's only for upgradeToRealUser
```

bpetetot commented on api/src/shared/domain/services/locale-service.js:
```diff
@@ -54,29 +52,18 @@ function coerceLanguage(language) {
-};
+function assertIsCanonicalLocaleName(locale) {
+  const canonicalLocaleName = Intl.getCanonicalLocales(locale)[0];
+  if (canonicalLocaleName != locale) throw new RangeError(`Non canonical locale: "${locale}"`);
```
La `RangeError` devrait être mappée dans le error manager, sinon on aura une erreur 500. On devrait avoir une erreur 400.

bpetetot commented on api/src/identity-access-management/application/oidc-provider/oidc-provider.controller.js:
```diff
@@ -53,9 +54,11 @@ async function authenticateOidcUser(request, h) {
+async function createUser(request, h, dependencies = { requestResponseUtils, localeService }) {
   const { identityProvider, authenticationKey } = request.deserializedPayload;
   const localeFromCookie = request.state?.locale;
+  if (localeFromCookie) dependencies.localeService.assertIsCanonicalLocaleName(localeFromCookie);
```
Si la locale du cookie, n'est pas supportée (par exemple, d'une façon ou d'une autre le cookie a été changé ou contient un mauvaise valeur). On empêchera l'utilisateur d'être créé, alors que la locale est 'facultative' ici. Je trouve cette différence de comportement est étrange.

Ça me semblerait plus résilient, si on mettait une locale par défaut, si on ne peut pas la parser.

lego-technix replied:
D’accord pour rendre résiliant. Mais si Joi+Hapi avait permis de faire une validation toute-prête de la valeur du cookie et qui aurait répondu `400` aux mauvaises valeurs, c’est ce que j’aurais préféré !

bpetetot reviewed (approved):
✅ Testé et review

pix-service commented:
:tada: This PR is included in version 5.178.0 :tada:

The release is available on [GitHub release](https://github.com/1024pix/pix/releases/tag/v5.178.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
